### ISSUE TITLE
Automate Release Process

### ISSUE THREAD

**Issue by artursouza** [dapr/dapr#3313]
## In what area(s)?

/area test-and-release
## Describe the proposal

Today, releasing Dapr requires the work of multiple people across multiple days (about 1 week). Automation should reduce the number of manual steps to a minimal.

Opportunities for automation in RC stage:
* Update helm chart for dapr/dapr
* Generate RCs for contrib, dapr, CLIs
* Validate RC installation and upgrade
* Release RCs for each SDK

Once RCs are all approved:
* Generate release notes automatically and allow manual editing by the team

Once release notes is finalized (approved), automate:
* Remove the RC prefix in all repos
* Update Helm chart in dapr/dapr
* Tag all repos
* Release each SDK
* Validate installation of released version

Ideally, there should also be a central workflow execution UI to visualize and follow the release steps.

**Comment by daixiang0** [comment]
What does "Remove the RC prefix in all repos" mean? Is it RC tag?

**Comment by artursouza** [comment]
it is the RC "suffix" - I fixed the description. It is because we label the releases with vX.Y.Z-RC.W and when we release, it must be vX.Y.Z - so, yes, it is the tag.

**Comment by daixiang0** [comment]
I still some RC tag from github UI.

btw, use tag rather than branch would be complex when backport, it need update all the time.

**Comment by artursouza** [comment]
We have a branch per release. Do you mean to have a branch per RC too?

**Comment by daixiang0** [comment]
with `git tag -l`, lots of RC tags exist, remove them is new here?

> We have a branch per release. Do you mean to have a branch per RC too?
No, my mistake, backport should only occur to vX.Y.Z(+1), not RC.

Also, consider the repo count, maybe split this into multiple sub tasks is better and easy to reach.

**Comment by wcs1only** [comment]
> with git tag -l, lots of RC tags exist, remove them is new here?

Is there any harm in having the tags? They are our source of truth for at what point release candidates were cut. We regularly refer back to them when investigating issues and understanding when they were introduced.

> Also, consider it is a epic task, maybe split this into multiple sub tasks is better and easy to reach:

> * a tool for RC/release generate
> * a tool for latest RC/release test including deploy, e2e, upgrade
> * a tool for RC/release note generate and do release
> * update helm charts in dapr/dapr and dapr/helmcharts

So we already have tools to do each one of these steps (except the helm chart updates), but they exist across different repos and have a dependency order that is manually done by the release team today. This issue is mainly about creating a global cross package workflow that can kick these steps off across different repos and do so in the correct order, with no intervention unless one of the steps fails. We will indeed be splitting this up into smaller subtasks once we get a chance to land on an overall strategy and prioritize which tasks need to be automated the most.

**Comment by wcs1only** [comment]
> > with git tag -l, lots of RC tags exist, remove them is new here?
> 
> Is there any harm in having the tags? They are our source of truth for at what point release candidates were cut. We regularly refer back to them when investigating issues and understanding when they were introduced.

Rereading this, and the description. I think we're just having a misunderstanding in terminology. By remove RC suffix, I think what @artursouza means is PRs like this one: https://github.com/dapr/python-sdk/pull/212

**Comment by artursouza** [comment]
Yes, "removing RC suffix" does not mean removing the tag itself but creating a new one without the suffix.

And yes, this issue is an Epic. We are using this issue as a parent issue.

**Comment by mikeee** [comment]
Proposal - https://github.com/dapr/proposals/pull/68
